### PR TITLE
[backup] BackupStorage trait

### PR DESCRIPTION
**Pull request by msmouse**
## Motivation

Talk AsyncRead / AsyncWrite instead of Stream<Type = Vec<u8>>.
Introduce the concept of BackupHandler.Talk AsyncRead / AsyncWrite instead of Stream<Type = Vec<u8>>.
Introduce the concept of BackupHandler.
### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

existing coverage

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/website, and link to your PR here.)

### CODE REVIEW

**Comment by wqfish** [comment]
Is https://github.com/libra/libra/pull/4054 part of this?

**Comment by msmouse** [comment]
> Is #4054 part of this?

Yes this stacks on that. Rebased now that it has landed.

File: storage/backup/backup-cli/src/storage/mod.rs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+pub type BackupHandleRef = str;
+pub type FileHandle = String;
+pub type FileHandleRef = str;
+
+#[async_trait]
+pub trait BackupStorage {
+    /// Hint that a bunch of files are gonna be created related to a backup identified by `name`,
+    /// which is unique to the content of the backup, i.e. it won't be the same name unless you are
+    /// backing up exactly the same thing.
+    /// Storage can choose to take actions like create a dedicated folder for do nothing.
```
Comment (wqfish) [inline storage/backup/backup-cli/src/storage/mod.rs:20]: is there a typo in "for do nothing"?

Comment (msmouse) [reply storage/backup/backup-cli/src/storage/mod.rs:20]: right.. *or* do nothing

**Pull request closed without merging**